blog/2021-06-16-stryker4s-sbt-performance.mdx

```diff
@@ -0,0 +1,156 @@
+
+1. `fullClasspath` location of all dependency JARs (including test-dependencies) and `.class` files of your compiled code.
+2. `loadedTestFrameworks` the configured test framework. Sbt detects a couple of frameworks by default, but if you use MUnit you might remember having to add `testFrameworks += new TestFramework("munit.Framework")` to your build.sbt before MUnit can run.
+3. `testGrouping` contains every test suite sbt discovers in your project.
+
+You can try running these yourself in a sbt console to see how Stryker4s looks at your project:
+
+```
+sbt:stryker4s-core> show fullClasspath
+[info] * Attributed(~/stryker4s/core/target/jvm-2.13/classes)
+[info] * Attributed(~/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-
+library/2.13.4/scala-library-2.13.4.jar)
+[info] * Attributed(~/.cache/coursier/v1/https/repo1.maven.org/maven2/com/github/pathikrit/
+better-files_2.13/3.9.1/better-files_2.13-3.9.1.jar)
+...
+sbt:stryker4s-core> show loadedTestFrameworks
+[info] Map(TestFramework(org.scalatest.tools.Framework, org.scalatest.tools.ScalaTestFramework) -> org
+.scalatest.tools.Framework@2b9d52d5)
+...
+sbt:stryker4s-core> show testGrouping
+[info] * (<default>,Vector(Test stryker4s.config.TestFilterTest : subclass(false, org.scalatest.Suite)
+, Test stryker4s.extension.FileExtensionsTest : subclass(false, org.scalatest.Suite), Test stryker4s.m
+utants.MutatorTest : subclass(false, org.scalatest.Suite), Test stryker4s.mutants.AddAllMutationsTest
+: subclass(false, org.scalatest.Suite), ...
+```
+
+We now know what tests are in your project. Let's have a look at running those tests:
+
+### 🧪 Meet: test interfaces
+
+This is something our friends in the NodeJS world are (rightfully) a little [jealous of](https://stryker-mutator.io/docs/stryker-js/plugins#test-runners) 😏. When you call `sbt test`, sbt will look for any registered test frameworks that implement the [test-interface](https://github.com/sbt/test-interface) (from `loadedTestFrameworks`). This is a small collection of interfaces that let a test framework tell sbt how to run tests in that framework. Because all test frameworks in sbt implement this interface, we have a single programmatic way to call any test framework. It doesn't matter if you're using ScalaTest, MUnit, Weaver-test or anything else. Using a bit of reflection, we can create an instance of the test framework runner and run tests:
+
+```scala
+val framework = classLoader.loadClass(testFrameworkClass).getConstructor().newInstance().asInstanceOf[Framework]
+val runner = framework.runner(args, remoteArgs, classLoader)
+
+val testTasks: Array[Task] = runner.tasks(testGrouping.taskDefs)
+
+testTasks.foreach(testTask => {
+  testTask.execute()
+})
+```
```

Comment:
I think a line here explaining how this test interface is used in Stryker4s would be helpful.